packages/build-tools/src/common/projectSources.ts

```diff
@@ -23,34 +24,74 @@ export async function prepareProjectSourcesAsync<TJob extends Job>(
       destinationDirectory
     );
   } else if (ctx.job.projectArchive.type === ArchiveSourceType.GIT) {
-    await shallowCloneRepositoryAsync(
-      ctx,
-      ctx.job.projectArchive.repositoryUrl,
-      ctx.job.projectArchive.gitRef,
-      destinationDirectory
-    );
+    await shallowCloneRepositoryAsync({
+      logger: ctx.logger,
+      archiveSource: ctx.job.projectArchive,
+      destinationDirectory,
+    });
   }
 }
 
-async function shallowCloneRepositoryAsync<TJob extends Job>(
-  ctx: BuildContext<TJob>,
-  projectRepoUrl: string,
-  gitRef: string,
-  destinationDirectory: string
-): Promise<void> {
+async function shallowCloneRepositoryAsync({
+  logger,
+  archiveSource,
+  destinationDirectory,
+}: {
+  logger: bunyan;
+  archiveSource: ArchiveSource & { type: ArchiveSourceType.GIT };
+  destinationDirectory: string;
+}): Promise<void> {
+  const { repositoryUrl } = archiveSource;
   try {
     await spawn('git', ['init'], { cwd: destinationDirectory });
-    await spawn('git', ['remote', 'add', 'origin', projectRepoUrl], { cwd: destinationDirectory });
-    await spawn('git', ['fetch', 'origin', '--depth', '1', gitRef], { cwd: destinationDirectory });
-    await spawn('git', ['checkout', gitRef], { cwd: destinationDirectory });
+    await spawn('git', ['remote', 'add', 'origin', repositoryUrl], { cwd: destinationDirectory });
+
+    let gitRef: string | null;
+    let gitCommitHash: string;
+    // If gitRef is provided, but gitCommitHash is not,
+    // we're handling a legacy case - gitRef is the commit hash.
+    // Otherwise we expect gitCommitHash to be present.
+    if (archiveSource.gitRef && !archiveSource.gitCommitHash) {
+      gitCommitHash = archiveSource.gitRef;
+      gitRef = null;
+    } else {
+      gitCommitHash = nullthrows(archiveSource.gitCommitHash);
+      gitRef = archiveSource.gitRef ?? null;
+    }
+
+    await spawn('git', ['fetch', 'origin', '--depth', '1', '--no-tags', gitCommitHash], {
+      cwd: destinationDirectory,
+    });
+
+    await spawn('git', ['checkout', gitCommitHash], { cwd: destinationDirectory });
+
+    // If we have a gitRef, we try to add it to the repo.
+    if (gitRef) {
+      const { name, type } = getStrippedBranchOrTagName(gitRef);
+      switch (type) {
+        // If the gitRef is for a tag, we add a lightweight tag to current commit.
+        case 'tag': {
+          await spawn('git', ['tag', name], { cwd: destinationDirectory });
+          break;
+        }
+        // gitRef for a branch may come as:
+        // - qualified ref (e.g. refs/heads/feature/add-icon), detected as "branch" for a push,
+        // - unqualified ref (e.g. feature/add-icon), detected as "other" for a pull request.
+        case 'branch':
+        case 'other': {
+          await spawn('git', ['checkout', '-b', name], { cwd: destinationDirectory });
+          break;
```

Comment:
Is it possible to pull/fetch a branch in a way so we don't need to create it locally 🤔

Reply:
Hm, so:
- we could do what GitHub is doing — pull a commit pointing to a remote branch, then create a new local branch pointing to the remote branch head — this feels overly complicated though
- we could do something slightly simpler — pull a commit pointing to a local branch (thus creating it via a Git ref spec like `git fetch +${commit}:${branch}`. This would create the branch locally behind the scenes. (I believe this solution is what you're suggesting?)

Do you think it would be better? At first I wasn't a fan of introducing the Git ref spec concept to here `+${commit}:${ref}`, but now after playing with it a bit more I can see the benefit of having only two commands:

- `git fetch +${commit}:${ref}`
- `git checkout ${ref || commit}`

I have just tested both commands work for all: qualified branch (`refs/heads/main`) and tag (`refs/tags/v1.0`) refs and an unqualified branch ref (`main`).

Reply:
both approaches seem good to me, I just wanted to understand why and how better